### PR TITLE
fix(icons-angular): build improvements

### LINKED ISSUE #3519: CI failing on icons-angular build step: Allocation failed - JavaScript heap out of memory
## Detailed description

> Describe in detail the issue you're having.

Our CI test runner is failing with the error: FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

I'm unable to reproduce this error locally, it likely has to do with just the CircleCI test runner

here is an example stack trace: https://circleci.com/gh/carbon-design-system/carbon/4254

### LINKED ISSUE #3522: icons-angular: memory issues causing circleci fail. build is currently deactivated.
icons-angular failing at build, likely due to memory issues with the extremely-large number if icons we have.

`packages/icons-angular/tasks/build.js` has been commented out to prevent `icons-angular` from building, until a fix is found.

ref issues:

- https://github.com/carbon-design-system/carbon/pull/3501#discussion_r306563648 
  - this is the PR which comments out the `icons-angular` build step
- https://github.com/carbon-design-system/carbon/issues/3519
  - this is another reporting of this error with some detail and specifics of stack trace

### PR DESCRIPTION
**Pull request by cal-smith**
- switches to bazel as the primary build system
- resolves OOM errors
- complies with APF for better compatibility with Ivy/NGCC

closes #3522 
closes #3519

### CODE REVIEW

**Comment by cal-smith** [comment]
Well, it seems like the workers are running out of resources and crashing out. So that's fun. In addition Bazel has some issues determining CPU and memory counts when running in a container so we have to guesstimate values that will make it happy (since the container may or may not get 2 full CPUs, and 4 full GB of memory)

The traditional angular compiler also hits a resource limit. It appears we've found the limit for Angular compilation on a 2CPU/4GB instance. FYI @joshblack

**Comment by joshblack** [comment]
@cal-smith any thoughts on how we should address this? If it'd be easier to just export a single entrypoint versus ones per-icon let me know, we've started transitioning over to this strategy with icons-react.

**Comment by cal-smith** [comment]
I'll spend some more time on this next week, but it seems like the size of the compile is simply too large for the resources available ... A single entrypoint doesn't seem feasible given the quantity of code we're looking to ship (the `icons-react` umd is `1.6mb` and there's a lot less code in there than in `icons-angular`) - and even still, I have a suspicion that we may hit a OOM situation. 

I'm not sure what the solution is (outside of simply giving the build more resources) but it's clear that _something_ is going to have to change.

**Comment by cal-smith** [comment]
So, an option I've been playing with is enabling RBE (Remote Build Execution) ... I set up a demo environment locally with one of my VMs and it seems to work quite nicely (Specifically buildbarn on docker-compose - https://github.com/buildbarn/bb-deployments/tree/master/docker-compose). 

At the moment, the options on the table are:
1. somehow reduce build complexity (a difficult task given the size of the compilation we're working with)
2. increase CircleCI resource allocation
3. enable RBE through some high resource server/kubernetes cluster (could also be used to help local builds)

File: packages/icons-angular/.bazelrc
Comment (alan-agius4) [inline packages/icons-angular/.bazelrc]: In CI I also suggest you to add

```
build --verbose_failures
```

Also, in the circle logs, I noticed that you are running bazel build in parallel with some other build tasks or I saw wrong? 

Also, probably you want to have a different rc for circle since locally you won’t want to limit the resources like you do on circle

Comment (cal-smith) [reply packages/icons-angular/.bazelrc]: Hi @alan-agius4,

> In CI I also suggest you to add

Will do 👍

> Also, in the circle logs, I noticed that you are running bazel build in parallel with some other build tasks or I saw wrong?

Yep, that's correct ... This repo is a monorepo for our whole design system ... We're not quite at the point where we're looking in to integrating Bazel everywhere, but it's come up before.

You can see from the build logs that even with _very_ conservative limits we're running out of memory, even after several hours worth of building.

I plan to try and reduce the number of modules we're building (by consolidating them where possible). The Angular compiler/metadata generation seems to be the bottleneck in terms of compilation time and resource consumption, so I'm not sure if simply reducing the number of modules will gain us anything.

Comment (alan-agius4) [reply packages/icons-angular/.bazelrc]: Yeah reducing the modules might help since you have limited CI resources.

Also I suggest you add a cache for bazel, so that after a green build bazel will only need to build the changes and not start from scratch.

See:  https://github.com/carbon-design-system/carbon/pull/3915/commits/f245a7832a5e5735f96645a662ac17fdea2c01cf

(Sry for being brief, but I am replaying from my mob)

**Comment by cal-smith** [comment]
Integrated the changes from @alan-agius4's #3915, we're still seeing Java crash out ... Next step is consolidating the icon components ... should cut the files we have to compile by 1/3 or 1/4

```
ERROR: /home/circleci/repo/packages/icons-angular/ts/video--filled/BUILD.bazel:4:1: Compiling Angular templates (ngc) //ts/video--filled:video-filled failed: Worker process quit or closed its stdin stream when we tried to send a WorkRequest:

---8<---8<--- Exception details ---8<---8<---
java.io.IOException: Stream closed
	at java.base/java.lang.ProcessBuilder$NullOutputStream.write(Unknown Source)
	at java.base/java.io.OutputStream.write(Unknown Source)
	at java.base/java.io.BufferedOutputStream.flushBuffer(Unknown Source)
	at java.base/java.io.BufferedOutputStream.write(Unknown Source)
	at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.doFlush(CodedOutputStream.java:3003)
	at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.flushIfNotAvailable(CodedOutputStream.java:2998)
	at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.writeUInt32NoTag(CodedOutputStream.java:2825)
	at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.writeTag(CodedOutputStream.java:2670)
	at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.writeBytes(CodedOutputStream.java:2723)
	at com.google.devtools.build.lib.worker.WorkerProtocol$Input.writeTo(WorkerProtocol.java:223)
	at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.writeMessageNoTag(CodedOutputStream.java:2800)
	at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.writeMessage(CodedOutputStream.java:2775)
	at com.google.devtools.build.lib.worker.WorkerProtocol$WorkRequest.writeTo(WorkerProtocol.java:1009)
	at com.google.protobuf.AbstractMessageLite.writeDelimitedTo(AbstractMessageLite.java:98)
	at com.google.devtools.build.lib.worker.WorkerSpawnRunner.execInWorker(WorkerSpawnRunner.java:330)
	at com.google.devtools.build.lib.worker.WorkerSpawnRunner.actuallyExec(WorkerSpawnRunner.java:172)
	at com.google.devtools.build.lib.worker.WorkerSpawnRunner.exec(WorkerSpawnRunner.java:121)
	at com.google.devtools.build.lib.exec.SpawnRunner.execAsync(SpawnRunner.java:225)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:123)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:88)
	at com.google.devtools.build.lib.actions.SpawnActionContext.beginExecution(SpawnActionContext.java:41)
	at com.google.devtools.build.lib.exec.ProxySpawnActionContext.beginExecution(ProxySpawnActionContext.java:60)
	at com.google.devtools.build.lib.actions.SpawnContinuation$1.execute(SpawnContinuation.java:80)
	at com.google.devtools.build.lib.analysis.actions.SpawnAction$SpawnActionContinuation.execute(SpawnAction.java:1344)
	at com.google.devtools.build.lib.analysis.actions.SpawnAction.beginExecution(SpawnAction.java:314)
	at com.google.devtools.build.lib.actions.Action.execute(Action.java:123)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$4.execute(SkyframeActionExecutor.java:841)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.continueAction(SkyframeActionExecutor.java:975)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:947)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:116)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:77)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:571)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:742)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:260)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:451)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:399)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
---8<---8<--- End of exception details ---8<---8<---

---8<---8<--- Start of log, file at /home/circleci/.cache/bazel/_bazel_circleci/1873fb996b5ba20f151d9e193011e19c/bazel-workers/worker-0-AngularTemplateCompile.log ---8<---8<---
(empty)
---8<---8<--- End of log ---8<---8<---
Target //ts:npm_package failed to build
INFO: Elapsed time: 1056.002s, Critical Path: 557.33s
INFO: 178 processes: 164 processwrapper-sandbox, 14 worker.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

**Comment by joshblack** [comment]
@cal-smith feel free to use GitHub Actions btw if that would help!

As you mentioned, feel free to provide a full list of icons and treat sizes as props if that helps bring the compilation size down.

**Comment by cal-smith** [comment]
Oh, Github Actions may be a decent idea 🤔 I'll see what the resource limits look like on that ...

Consolidating the files seems to have helped a bit, but yeah it seems like the strategy may be to treat the size as an input and just dynamically swap that way - same number of total files as right now (since I've consolidated the modules) but way less total code.

**Comment by cal-smith** [comment]
I'm not sure the stale warning is particularly applicable in this case @joshblack. Apologies that this has taken so long, but I do have commitments beyond this (massive) PR.

**Comment by joshblack** [comment]
Hey @cal-smith! 👋 We're just trying to reduce the amount of PRs open to ones that are active, feel free to open/close whenever the timing works out! It's just to help our review process. Let me know if you want to chat on Slack about this!

**Comment by cal-smith** [comment]
GET HYPE

🎉 CIRCLE IS PASSING 🎉 

Now I just need to build locally and check that the compiled output works with Ivy and we'll be good to go

I'll also clean up the `package.json`s, and prune the yarn mirror

**Comment by cal-smith** [comment]
You can remove that stale warning now @joshblack 😉

**Comment by cal-smith** [comment]
Hmm ... We seem to be hitting https://github.com/angular/angular/issues/32061 

image

So, we can ship this, but that's definitely a blocker for Ivy compat ... we _need_ to be able to apply directives to SVGs

(CC @alan-agius4 in case you have any insight)

**Comment by cal-smith** [comment]
FYI @joshblack to accommodate some of the changes, I've switched to building to a `dist` folder ... is there something I should change to have `dist` be the target for publishing?

**Comment by cal-smith** [comment]
Bonus hype 🎉 

This is no longer a breaking change 😁 the old `lib` and `umd` folders are built as part of a separate "compat" build until users switch off to the new style modules. We'll still drop the compat build with v11, but this way users aren't waiting until then to pick up changes and start migrating.

File: package.json
Comment (vpicone) [inline package.json:17]: modifying the primary build script for a single package is not a good idea, if we do this for every package it would be insane.

**Review by vpicone — DISMISSED** [review]
I don't know enough about angular to review the framework code, my only feedback is to the modifications to the root `package.json`

File: package.json
Comment (cal-smith) [reply package.json:17]: Totally open to suggestions 👍 How would you recommend we approach this? Keep in mind we've hit both the CircleCI and node memory limit at various points. The node limit seems to be under control, but I worry that in the lerna streaming build we may exceed the CircleCI limit.

I'll push a test commit to see if we run into any issues.

Comment (cal-smith) [reply package.json:17]: Yep, so seems like under the streaming build we do indeed hit the CircleCI resource limit: `@carbon/icons-angular: error Command failed with exit code 137.` where `137` usually indicates a resource error (i.e. out of memory).

As much as I appreciate that modifying the primary build script isn't ideal, we do have a fairly exceptional circumstance on our hands - unless there's some other way to run this particular build in isolation.

File: packages/icons-angular/package.json
Comment (asudoh) [inline packages/icons-angular/package.json:11]: Is `files` covered by somewhere else?

File: packages/icons-angular/src/tsconfig-build.json
Comment (asudoh) [inline packages/icons-angular/src/tsconfig-build.json]: Do we mean to remove this line...?

**Review by asudoh — APPROVED** [review]
LGTM 👍 basically, just a couple of questions - Thanks @cal-smith for completing this!! 🎉

File: packages/icons-angular/package.json
Comment (cal-smith) [reply packages/icons-angular/package.json:11]: Yep 👍 `ng-packagr` handles generating the final `package.json`

File: packages/icons-angular/src/tsconfig-build.json
Comment (cal-smith) [reply packages/icons-angular/src/tsconfig-build.json]: We should be safe to do that yeah

File: packages/icons-angular/src/story-templates.js
Comment (emyarod) [inline packages/icons-angular/src/story-templates.js]: is this intentional?

**Review by emyarod — COMMENTED** [review]
I don't have an alternative approach to the build script changes but I'm ok with it as a stop gap due to circle ci limitations

File: packages/icons-angular/src/story-templates.js
Comment (cal-smith) [reply packages/icons-angular/src/story-templates.js]: It was ... though we could probably replace it with a `reduce` 🤔

Comment (emyarod) [reply packages/icons-angular/src/story-templates.js]: oh I was referring more to the name but a `reduce` would work here as well yeah, either approach is fine

Comment (cal-smith) [reply packages/icons-angular/src/story-templates.js]: Oh wow, I totally didn't see that 😅 yeah that should definitely be _generate_ - but I'll probably swap it out for a `reduce` anyway

Comment (cal-smith) [reply packages/icons-angular/src/story-templates.js]: done!

**Comment by cal-smith** [comment]
BTW Lerna adds a `directory` key to `publishConfig` to allow individual packages to configure where they publish from - https://github.com/lerna/lerna/tree/master/commands/publish#publishconfigdirectory

So if I understand the publishing procedure, that should be it for publishing changes (@joshblack true?)

**Review by dakahn — APPROVED** [review]

**Comment by asudoh** [comment]
@vpicone Would you agree [@cal-smith's latest comment wrt Angular-specific build script in root `package.json`](https://github.com/carbon-design-system/carbon/pull/3798/files#r330237829)? Thanks!

**Comment by vpicone** [comment]
@asudoh I'm not familiar with why this package in particular's build script is busting the circle ci limitations, but not comfortable with modifying the primary build script as a one-off work around. 

I think we need to find a way around this if possible. Maybe it's time we open a ticket with Circle CI to increase our [resource_class](https://circleci.com/docs/2.0/configuration-reference/#resource_class)

File: packages/icons-angular/src/templates-compat.js
Comment (joshblack) [inline packages/icons-angular/src/templates-compat.js:184]: Not sure if this has come up, but feel free to only use one icon in the storybook instead of generating ones for each one!

File: package.json
Comment (joshblack) [reply package.json:17]: I think the preference would be to pull things out of the monorepo if we have a situation like this versus slow down the whole, if that makes sense.

Definitely agreed, this is a crazy exceptional situation that we're still dealing with on the react side. Just so many assets to ship.

Comment (cal-smith) [reply package.json:17]: > I think the preference would be to pull things out of the monorepo if we have a situation like this versus slow down the whole, if that makes sense.

Sure ... My only concern would be with syncing updates - I don't want this package to lag behind.

Some alternatives:

- we can look at adding a Github action for this package and excluding it from the local build (The build times are kinda excessive) ... At least there we can be sure we're synced up release wise.
- pulling all the icon packages out into a separate repo - given that the react and vue build will also continue to grow.
- pulling `icons-angular` into `carbon-components-angular` (this gets a bit muddy when we consider the monorepo issue ... let's sync up on that)

I'll take a look at adding an action for this ... but let me know if you have any other ideas 👍

**Comment by cal-smith** [comment]
> Maybe it's time we open a ticket with Circle CI to increase our resource_class

That's definitely the most straightforward solution. I'd also suggest excluding the icon libraries from the local builds - Angular takes a minute or two, and React and Vue aren't super quick either.

As I mentioned above I'll look into adding a Github action that targets this package.

In the short term this is a solution that does get us back in a state where we can ship Angular icon components. I would vote to ship this (imperfect as it is), and immediately begin looking at ways we  can solve the core issue.

**Comment by asudoh** [comment]
As @cal-smith said, looks that we should go with the workaround in the interim and continue thinking about alternatives. Not certain if our CircleUI usage supports `resource_class`, but it may be one option. @vpicone?

**Comment by asudoh** [comment]
Merging, any maintainers feel free to back it out if any of you see an actual problem. Thanks!

**Comment by joshblack** [comment]
@asudoh can you revert this PR? There were several comments that we should have resolved before merging.

**Comment by vpicone** [comment]
@asudoh did we not want to at least talk about it first?

**Comment by vpicone** [comment]
Not sure we needed to merge this without even using yarn workspace commands? `cd`ing into an individual package from the primary build script is...ill-advised. I had some other options once the most direct proposals were attempted.

**Pull request merged**
